### ISSUE TITLE
Adopt typescript type definitions for F7 Core?

### ISSUE THREAD

**Issue by JasonKleban** [framework7io/framework7#2616]
I've typescript-annotated most of the F7 Core at https://github.com/JasonKleban/Framework7.d.ts.  It isn't 100% yet but it's a good start that the community can help refine.  At some point, adoption of those type definitions into the F7 distribution would be cleaner for F7 developers who wish to take advantage of type-safety and intellisense/autocomplete in supporting editors.

There is only one file for each of [F7 Core](https://github.com/JasonKleban/Framework7.d.ts/blob/master/framework7.d.ts), [Template7](https://github.com/JasonKleban/Framework7.d.ts/blob/master/template7.d.ts), and [Dom7](https://github.com/JasonKleban/Framework7.d.ts/blob/master/dom7.d.ts) (those other two would be distributed with their own official distributions too).

What is your appetite/concerns for adopting those definitions into this codebase?

**Comment by nolimits4web** [comment]
Hey @JasonKleban 

I'm for it with both hands. I think now yes, it can be very useful for developers. For better maintenance, i think it would be better if we just put same `.d.ts` file to every component and module that tells how it extends class/app instance how you already did in your main F7 Core `.d.ts` file. And then we can generate main framework7.d.ts file automatically in the same way like `.esm` module generated which basically will be the same amount if imports/exports. What do you think?

**Comment by JasonKleban** [comment]
I tried a few ways to split up the files.  There's a few options to strive for:  1) Single file .d.ts 2) Multiple .d.ts files with native uncompiled support among the multiple files somehow importing the parts that were chosen for the bundle 3) Multiple .d.ts files combined as part of a typescript-native build routine 4) Multiple .d.ts files with custom build step concatenation.

1) Is the default
2) Ideal, since you don't want to change your development technique or add build steps, but I'm doubting that it's possible to have multiple .d.ts files independent by module and ergonomic...
3) Requires typescript as a build tool, if possible at all.
4) Maybe, but this option sucks IMO because we likely have to develop the typescript definitions without native typescript tooling support.

Calling @bencompton for thoughts.

Regarding 2).. Since each module can add functionality to the Framework7 interface, this isn't as simple as importing.  What we need is like an `include ...` statement which I'm not aware of.  There's `/// <references path="..." />`  but I tried that and it doesn't work like an include.  I tried it with Dialog's stuff as a good example.

Regarding 4) ... Yes, it's a several-thousand line file, annoying to maintain.  But it's only a dev dependency, so minimal bundling isn't so important for consumers.  And even if a custom f7 bundle doesn't include a certain component, a very shallow and early tests will reveal that this dependency is missing.

So all things considered, I prefer option 1 of keeping it as one file where even custom F7 bundles include all component definitions.

**Comment by nolimits4web** [comment]
Hey @JasonKleban i spent a while today trying to figure out and find a way to make splitting work, wasn't to strait forward as i'm not too good in TypeScript. So here is the main d.ts file:
https://github.com/framework7io/framework7/blob/dts/src/core/framework7.d.ts
```
import Framework7 from './components/app/app-class';
import './components/accordion/accordion';
import './components/popup/popup';

export default Framework7;
```

Main declarations at:
https://github.com/framework7io/framework7/blob/dts/src/core/components/app/app-class.d.ts
```
export interface Framework7Class<Events> {
  ...
}

export interface Framework7EventsClass<Events> {
  ...
}

export interface Framework7Params {
  ...
}

export interface Framework7Plugin {
  ...
}

export interface Framework7AppEvents {
  ...
}

interface Framework7 extends Framework7Class<Events> {
  ...
}

declare class Framework7 implements Framework7 {
  constructor(parameters?: Framework7Params);

  static use(plugin : Framework7Plugin) : void;
}

export default Framework7;
```

And each component has its own d.ts with just extending main interfaces, e.g. Accordion:
https://github.com/framework7io/framework7/blob/dts/src/core/components/accordion/accordion.d.ts
```
import Framework7 from '../app/app-class';
import { CSSSelector } from '../app/app-class';

declare module '../app/app-class' {
  interface Framework7Class {
    accordion: {
      /** open specified accordion item */
      open(el : HTMLElement | CSSSelector) : void

      /** close specified accordion item */
      close(el : HTMLElement | CSSSelector) : void

      /** toggle specified accordion item */
      toggle(el : HTMLElement | CSSSelector) : void
    }
  }
  interface Framework7Params {
    //
  }
  interface Framework7AppEvents {
    /** Event will be triggered when accordion content starts its opening animation */
    accordionOpen : (el : HTMLElement | CSSSelector) => void

    /** Event will be triggered after accordion content completes its opening animation */
    accordionOpened : (el : HTMLElement | CSSSelector) => void

    /** Event will be triggered when accordion content starts its closing animation */
    accordionClose : (el : HTMLElement | CSSSelector) => void

    /** Event will be triggered after accordion content completes its closing animation */
    accordionClosed : (el : HTMLElement | CSSSelector) => void
  }
}
```
A bit more complex case with popup:
https://github.com/framework7io/framework7/blob/dts/src/core/components/popup/popup.d.ts

From first tests it works correct, it correctly extends Framework7 constructor parameters, instance methods, app events, etc. So can we have some problems with such structure? Not also sure do we need `DomEvents` extensions?

**Comment by nolimits4web** [comment]
I pushed it all to `dts` branch here

**Comment by nolimits4web** [comment]
TypeScript definitions landed in 3.3.0
